### PR TITLE
Feat/draft v standard

### PR DESCRIPTION
**Pull request by Stebalien**

### CODE REVIEW

**Review by mikeal — DISMISSED** [review]
We need some way of noting that this is a breaking change for any dependent projects.

Especially since this column is not being appended to the end, this is going to change/break the representation that comes back from many CSV parsers. It’s pretty common to throw out the header line and just parse each line into an array rather than parsing each line into a map using the header values as key names.

**Comment by Stebalien** [comment]
I agree, that's why the tests are failing. But I'm not sure how to actually _do_ that.

**Comment by mikeal** [comment]
Is anyone pulling the latest master automatically or are all the known libraries keeping their own cached version?

**Comment by Stebalien** [comment]
go-multihash is pulling into a submodule but (a) that's locked against a specific commit and (b) we're only using it for tests.

@vmx?

File: README.md
Comment (vmx) [inline README.md]: What about renaming "standard" to "permanent" like the [URI Schemes Registry](https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml) is doing it. That would make clear that having a standard, might still not be enough to get the permanent assignment here.

I think "draft" is a good name.

**Comment by vmx** [comment]
In the JS world there is a script that takes the file that generates JS code. It's a manual process.
In Rust, there is not even a script, there it's totally manually.

**Comment by rvagg** [comment]
So, merging this and breaking people's stuff will tell them that they should probably be relying on the header to know what the columns are rather than hard-wiring it in. So I'm +1 on merging and taking care of the consumers that we control and adding in documentation that if you're consuming the table _you must use the header to determine column contents_.

**Comment by ribasushi** [comment]
You know - we could just switch to a slightly more extensible format: JSON, YAML, whatever: something that allows us to grow the metadata without rebreaking the world every time.

We can also keep the .csv around by siply having a step that regenerates it from the now-canonical master-spec.

I would volunteer to implement this, together with the necessary CI hooks, provided there is buy in to go that route.

**Comment by rvagg** [comment]
XML please, lashings of it, with namespaces too, and a schema while you're at it!

+1 to a JSON version as the primary. We could even just publish it to npm as is for consumption by js-multicodec and js-multihash (maybe). Let's see what @vmx says.

**Comment by vmx** [comment]
For this case I don't like JSON for the lack of being able to add comments, but so is CSV => I think we should go with JSON as canonical representation, and generate CSV out of it (as CSV is really so much easier to read/look things up).

**Comment by mikeal** [comment]
Why don’t we just use a markdown table and write a simple script that bulids a .csv and .json file on every checking using GitHub Actions?

**Comment by rvagg** [comment]
> just use a markdown table

Because it's not that much different from a CSV then, we haven't solved very much. We may as well have an "input CSV" an a "consume-me CSV" if we head down that route. The only thing a Markdown table really buys us is the ability to embed it in a large Markdown doc (the table is getting large enough that this probably isn't something we need to be doing) or to discourage it from being consumed as the primary source (documentation can do that job, as can naming).

@vmx is there much more than we'd need in terms of comments than we could fit into a `"comments"` field for each entry? That can be arbitrary text with comments about anything relating to an entry.

**Comment by vmx** [comment]
> @vmx is there much more than we'd need in terms of comments than we could fit into a `"comments"` field for each entry?

Good point. I think that's enough.

**Comment by Stebalien** [comment]
I don't have the time to refactor how we store multicodecs but I really want this merged so we can more easily merge new multicodec registrations.

File: table.csv
Comment (vmx) [inline table.csv:96]: I would consider dag-json being permanent. But we can also discuss this in a seperate PR.

File: table.csv
Comment (vmx) [inline table.csv:102]: Would it make sense to call `ipfs` draft and `p2p` to be permanent?

File: table.csv
Comment (vmx) [inline table.csv:195]: Is intentionally Blake2**b** permanent and Blake2**s** draft?

**Comment by vmx** [comment]
@Stebalien you mention that having "draft" and "permanent" would make merging new multicodecs easier. My impression is that this wasn't really the problem so far when we didn't merge things. I think the problem is that we miss a clear definition what multicodec codes (and their categories) are about. It boils down to: Is it just a random table with numbers in it, or do they have a certain meaning.

File: table.csv
Comment (Stebalien) [reply table.csv:102]: Good point.

File: table.csv
Comment (Stebalien) [reply table.csv:195]: I'm not aware of anyone using the blake2s codec in practice.

**Comment by Stebalien** [comment]
> It boils down to: Is it just a random table with numbers in it, or do they have a certain meaning.

Well, _that_ part usually boils down to "are they using IPLD right". But draft v permanent would make me feel better about adding new codes.

**Review by vmx — APPROVED** [review]
Before merging I'd like to also have the approval from @rvagg, as he also spent quite a lot of time on all this.

File: table.csv
Comment (rvagg) [reply table.csv:195]: you could also make this argument for most of the blake2b hashes, `blake2b-256` is the only one in _wide_ use (because of Filecoin)

**Review by rvagg — APPROVED** [review]
I'm fine with the column, its the first-pass of filling it in that I think has issues. We can resolve them later but these are the most important discrepancies from my perspective:

* only one of the blake2b hashes has widespread usage yet they are all permanent here
* sha2-256-trunc254-padded and poseidon-bls12_381-a2-fc1 are the CID pairs for fil-commitment-unsealed and fil-commitment-sealed respectively. The former are draft but the latter are permanent.

Second-level concerns I have are:

* I think the bitcoin* and eth* codecs should be permanent, they are being used in the wild, outside of PL, and as more than just examples (dbl-sha2-256 would come along with bitcoin*)
* cbor and json would be nice to have as permanent, although I can't point to active non-example uses of them
* dag-cose is getting more use in the wild I think, but we could wait for input from those users

**Comment by Stebalien** [comment]
> only one of the blake2b hashes has widespread usage yet they are all permanent here

My concern is that poking holes in this range could be troublesome if someone actually _needs_ one of the others. But then again, we might as well stick to our policy.

> sha2-256-trunc254-padded and poseidon-bls12_381-a2-fc1 are the CID pairs for fil-commitment-unsealed and fil-commitment-sealed respectively. The former are draft but the latter are permanent.

Will fix.

> I think the bitcoin* and eth* codecs should be permanent, they are being used in the wild, outside of PL, and as more than just examples (dbl-sha2-256 would come along with bitcoin*)

Fair enough.

> cbor and json would be nice to have as permanent, although I can't point to active non-example uses of them
> dag-cose is getting more use in the wild I think, but we could wait for input from those users

SGTM.

**Comment by Stebalien** [comment]
> dag-cose is getting more use in the wild I think, but we could wait for input from those users

Jose or Cose or both?

**Comment by Stebalien** [comment]
@rvagg I've fixed everything but jose/cose.

**Review by rvagg — APPROVED** [review]
leave cose/jose as draft for now

👍

**Pull request merged**
